### ISSUE TITLE
Preflight check issue before returning to e-shop

### ISSUE THREAD

**Issue by AndreasLoukakis** [VivaPayments/API#521]
If:
-the user is redirected from an e-shop to viva payments with sourceCode param ( https://www.vivapayments.com/web/newtransaction.aspx?ref=xxx&SourceCode=xx ) which means that after payment there should be a redirect back to the e-shop)
-the user selects viva wallet for his payment (not credit card)
-the e-shop has 301 redirects either from www to non-www or vise versa 
-the user pays the amount from his wallet...

..there is an unhandled issue as I noticed in vivapayments page console, where a preflight check to the merchant domain fails*. There is no notification to the user, who will most probably think his payment is not made / accepted and might try again, still with no message.
*the preflight check is to www.merchantsite.com which redirects to non-www merchantsite.com with a 301

I can understand the reason for the preflight check and why a redirect might not be acceptable but the way it's being handled is not really helping anything.

First of all, if the user completes a payment, he should know. Always. No matter what.
Secondly, if the redirect is from a www to non-www for the same domain, you should accept it. After all you do accept it if the payment is via card and not viva wallet.

Please do make the necessary adjustments to your api or advice on how to alternatively handle these cases.

<img width="1075" alt="screen shot 2018-04-29 at 18 57 45" src="https://user-images.githubusercontent.com/794135/39408688-b91869e8-4be2-11e8-8a4b-770b6a47e8fc.png">

**Comment by dpnevmatikos** [comment]
Hi @AndreasLoukakis ,
First of all I would like to apologize for the delayed response and thank you for taking the time
to provide all the details which really helped us to properly debug the issue.

From the in-house research, we discovered that indeed when the preflight request failed, checkout
was not redirecting back to the merchant's website. The issue caused had nothing to do with
the domain's type (naked or canonical). It's main cause was an HTTP request header (X-Requested-With) which made the browser to treat the request as an [un-safe request](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) (check section "Preflighted requests") and thus making a preflight request.



>First of all, if the user completes a payment, he should know. Always. No matter what.

You are right about that. From a user's standpoint my guess is that most users, seeing that nothing happens on the checkout page, would refresh the page to re-try the charge. In that case and because the payment would be successfully completed, they would be notified that the payment has been completed, redirecting them to the merchant's website. It's important to mention that no transactions were affected because of this issue.

The issue has been resolved, and both production and demo environment have been updated with the patch.

Again, apologies for the delayed response and for any inconvenience caused.
